pkg/controller/buildrun/buildrun_controller.go

```diff
@@ -278,6 +278,15 @@ func (r *ReconcileBuildRun) Reconcile(request reconcile.Request) (reconcile.Resu
 				buildRun.Labels = make(map[string]string)
 			}
 
+			// Set OwnerReference for Build and BuildRun only when build.build.dev/build-run-deletion is set "true"
+			if build.GetAnnotations()[buildv1alpha1.AnnotationBuildRunDeletion] == "true"{
+				if err := r.setOwnerReferenceFunc(build, buildRun, r.scheme); err != nil {
+					build.Status.Reason = err.Error()
+					updateErr := r.client.Status().Update(ctx, build)
+					return reconcile.Result{}, handleError("Failed to set OwnerReference for Build and BuildRun", err, updateErr)
+				}
+			}
```

Comment:
We are here in the buildrun controller and we are setting an owner reference for the buildrun. The build is not touched by this as far as I understood it. Why should the build's status be changed here?

Also, the `setOwnerReferenceFunc` call causes an update on the buildrun. This is only saved if the labels that are set a few lines later are not yet up-to-date. This should work but there is a chance that this does not work (if the buildRun is already created with these two labels correctly set). When correcting this, please make sure to continue to save labels and owner reference as one step and not introduce another save.

Reply:
@SaschaSchwarze0  Thank you for pointing this out. Sometimes it is a hard to uncover my own blind spot. I need to reconsider the logic.

Reply:
> The build is not touched by this as far as I understood it. Why should the build's status be changed here?

Because we are saying to the BUILD that now is owning a BUILDRUN. Similar to when we set ownerships, from a BUILDRUN to a TASKRUN, we update the BUILDRUN status if the ownership setting fails.

> Also, the setOwnerReferenceFunc call causes an update on the buildrun. This is only saved if the labels that are set a few lines later are not yet up-to-date. This should work but there is a chance that this does not work (if the buildRun is already created with these two labels correctly set).

I think is a valid edge case. @routerhan will be working on it.

Reply:
>> The build is not touched by this as far as I understood it. Why should the build's status be changed here?
>
> Because we are saying to the BUILD that now is owning a BUILDRUN. Similar to when we set ownerships, from a BUILDRUN to a TASKRUN, we update the BUILDRUN status if the ownership setting fails.

I think there is a slight difference between these two scenarios.

The TaskRun is only an implementation detail of the BuildRun and its owner reference is established when the our code is creating the TaskRun. Any failure that we notice while creating the TaskRun (including establishing the owner reference) makes sense to be reported on the BuildRun because that artifact is broken. (You would not be able to report it on the TaskRun anyway because that is not yet existing.)

The scenario is different for Build to BuildRun. The Build is created by the user and exists and the BuildRun is created by the user and exists. After the BuildRun has been created, our controller code wants to go ahead and modify it to add the owner reference to it (the Build is untouched). I am not sure in which case this can throw an error at all, but if it would, it is still our buildrun controller that tries to do something on a buildrun. If a failure happens, we can do imo two things:

1. Log and return the error to cause a retry of the reconciliation.
2. Mark the BuildRun as failed. The user will probably be not knowing what to do with the error, but he might just trigger a new BuildRun that might succeed.

Setting the Build's reason would mean that it is `Registered=True` and have the `Reason=err.Error()`. One could now argue about whether `Registered` could be set to `False` to completely break it, but, in any case, I do not see what the user would be doing with this error there.

Reply:
Thanks, good feedback, however I dont think we should failed the BuildRun or trigger more reconciles for the BuildRun, I think we should just update the Build `Reason`. Let me explain:

1. The "error" that could happen during the ownership definition is only related to(_[func](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go#L57-L89)_):
 - owned object is already owned
 - owner and owned object are not on the same namespace
 - owner groupversion cannot be found 

The above is most probably never gonna happen because:
- buildrun never gets own by anyone
- both build and buildrun are always on the same namespace
- build have a well defined groupversion

the above error most probably will never happen, and if it happens I dont think it worth failing the buildrun. Also, if I got this right, and the above three points could lead to a failure, then I think if we reconcile again, further reconciliations will always fail, therefore reconciling again doesnt make too much sense.

2. This logic despite being worked on the build-controller, have its trigger on the Build(_via a particular annotation_), so if the above error(_my point 1._) happens, this should be reflected in the Build, under the `Reason` column, without altering the value of `Registered` column. 

3. At the end of the day, if the ownership setting fails, a user should see this reflected on the Build Status columns, and should be able to understand that the BuildRun requires a manual intervention(_if a deletion is required_). Therefore a meaningful error needs to be carefully added.


For 3. we need nice unit tests, that test the scenarios when ownership fails and that we can see how the error will be displayed to the user via the Build.

Reply:
@SaschaSchwarze0 let us know what do you think, so that @routerhan could work on the right direction next Tuesday.

Reply:
Given the error will never happen, it is probably not worth spending more effort in the discussion then.

I am not convinced by all your arguments.

Not reconciling again as the theoretic error cannot come again, I agree based on the list of possible causes.

The annotation is on the build, that is correct. But, that does not by itself trigger the code in the buildrun controller. This runs due to the creation of the buildrun - which is the trigger from a user's perspective. A user who just created a build run would likely be checking it (= the build run) as a next step and would wonder that it continues to have an empty status. Looking for the cause of this at the build is imo not meeting his expectation.

Anyway, as we see no reason for the error situation to ever happen, I can live with any error handling as long as it includes a `ctxlog.Error` call.